### PR TITLE
test/cls_2pc_queue: prevent list+remove race between consumers

### PR DESCRIPTION
**Pull request by yuvalif**
* make sure that the queue full condition is covered
* add cls debugging to test

Fixes: https://tracker.ceph.com/issues/67229

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by yuvalif** [comment]
teuthology passing: https://pulpito.ceph.com/yuvalif-2024-07-29_18:53:13-rgw:verify-wip-yuval-67229-distro-default-smithi/

**Comment by yuvalif** [comment]
could not see the CLS debug messages in the osd logs on the remote machines:
* http://qa-proxy.ceph.com/teuthology/yuvalif-2024-07-29_18:53:13-rgw:verify-wip-yuval-67229-distro-default-smithi/7824592/remote/smithi174/log/
* http://qa-proxy.ceph.com/teuthology/yuvalif-2024-07-29_18:53:13-rgw:verify-wip-yuval-67229-distro-default-smithi/7824592/remote/smithi098/log/

File: qa/suites/rgw/verify/tasks/cls.yaml [hunk 1]
```diff
@@ -1,3 +1,8 @@
+overrides:
+  ceph:
+    conf:
+      osd:
+        debug cls: 20
```
Comment (cbodley) [inline qa/suites/rgw/verify/tasks/cls.yaml:5]: sorry, its `debug objclass`

> common/subsys.h:SUBSYS(objclass, 0, 5)

**Comment by yuvalif** [comment]
another teuthology run: https://pulpito.ceph.com/yuvalif-2024-07-30_18:05:24-rgw:verify-wip-yuval-67229-distro-default-smithi/
* the cls_2pc_queue tests are passing
* failure in s3select `test_schema_definition` is unrelated

cls logs are seen. e.g. http://qa-proxy.ceph.com/teuthology/yuvalif-2024-07-30_18:05:24-rgw:verify-wip-yuval-67229-distro-default-smithi/7826705/remote/smithi116/log/ceph-osd.2.log.gz

**Comment by yuvalif** [comment]
jenkins test api

**Comment by yuvalif** [comment]
jenkins test make check

**Comment by yuvalif** [comment]
jenkins test make check

File: src/test/cls_2pc_queue/test_cls_2pc_queue.cc [hunk 2]
```diff
@@ -965,33 +965,36 @@ TEST_F(TestCls2PCQueue, MultiProducerConsumer)
...
           bool truncated = true;
           std::string end_marker;
           std::vector<cls_queue_entry> entries;
           while (producer_count > 0 || truncated) {
+            if (!retry_happened) {
+              // queue was never full, let it fill
+              std::this_thread::sleep_for(std::chrono::milliseconds(100));
+              continue;
+            }
+            std::lock_guard lock(consumer_mutex);
```
Comment (cbodley) [inline src/test/cls_2pc_queue/test_cls_2pc_queue.cc:982]: why add a mutex over the calls to cls_2pc_queue_list_entries+cls_2pc_queue_remove_entries?

don't we want the consumers to race? if not, maybe just spawn a single consumer thread

Comment (yuvalif) [reply src/test/cls_2pc_queue/test_cls_2pc_queue.cc:982]: you are right. no point in testing multiple consumers if we serialize their execution.
in the rgw i use a cls lock to make sure there is no race between consumers.

Comment (yuvalif) [reply src/test/cls_2pc_queue/test_cls_2pc_queue.cc]: kepts multi threaded readers and a change the deletion to a single thread

**Comment by yuvalif** [comment]
teuthology: http://qa-proxy.ceph.com/teuthology/yuvalif-2024-08-05_13:10:00-rgw:verify-wip-yuval-67229-distro-default-smithi/7836950/teuthology.log

all relevant tests are passing

File: qa/suites/rgw/verify/tasks/cls.yaml [hunk 3]
```diff
@@ -1,3 +1,8 @@
+overrides:
+  ceph:
+    conf:
+      osd:
+        debug objclass: 20
```
Comment (cbodley) [inline qa/suites/rgw/verify/tasks/cls.yaml:5]: does this make the osd log files significantly larger? if so we may not to merge this part

**Review by cbodley — APPROVED** [review]
looks ok, but i've lost track of the purpose of this test case

we know that cls write ops are atomic. what more are we trying to prove with this test?

File: qa/suites/rgw/verify/tasks/cls.yaml (hunk 3 above)
Comment (yuvalif) [reply qa/suites/rgw/verify/tasks/cls.yaml:5]: what is the point in cls tests if we cannot use logs to figure out the issues?

**Comment by yuvalif** [comment]
> looks ok, but i've lost track of the purpose of this test case
> 
> we know that cls write ops are atomic. what more are we trying to prove with this test?

there are quite a bit of tests that simulate multiple clients operating at the same time.
if you think there could never be an issue with concurrency in cls we can remove all of them.

**Comment by yuvalif** [comment]
jenkins test make check

**Comment by yuvalif** [comment]
jenkins test api

**Comment by yuvalif** [comment]
@cbodley are you ok with merging as-is?
* keep cls log level 20 for cls tests
* keep the multithreaded tests

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
https://jenkins.ceph.com/job/ceph-pull-requests/141479/

> npm ERR! URL: https://download.cypress.io/desktop/12.17.4?platform=linux&arch=x64
npm ERR! Error: Corrupted download

**Comment by cbodley** [comment]
jenkins test make check

**Pull request merged**